QA Claude here 🔬 — follow-up to my heads-up above, now that the fix exists.

The `DoBlockMaxWand` long-query hang fix landed on A3 (weaviate/weaviate#11773) as `407aa75cc6` — but it's **only on that branch**. A4–A7 are still stacked on the pre-fix A3, so none of them carry it. I dropped the committed regression test (`search_segment_longquery_test.go`) onto this PR's tip (`18875a659e`) and it **fails**:

```
search_segment_longquery_test.go:105: DoBlockMaxWand did not terminate within 20s
  on a 200-term query (pivot not advancing)
--- FAIL: TestBlockMaxWandLongQueryTerminates (22.28s)
```

So the hang is live on the merge-tip of the whole stack — A5's `AdvanceAtLeastShallow` changes do **not** incidentally mask it — and the regression test is absent from all four PRs, so their CI can't catch it either.

| branch | tip | fix `407aa75cc6` | long-query test |
|---|---|---|---|
| bmw-p3-wand-loop (#11773) | `407aa75cc6` | ✅ present | ✅ green |
| bmw-p5-deferred-tombstone (#11774) | `f3ef5b4916` | ❌ absent | ❌ absent |
| bmw-p8-block-scan (#11775) | `6ab77c56c4` | ❌ absent | ❌ absent |
| bmw-p6-round6-setup (#11776) | `ecee68c759` | ❌ absent | ❌ absent |
| bmw-p4-round7-sort-elision (#11777) | `18875a659e` | ❌ absent | ❌ absent |

Action: before this stack merges, A4→A7 need re-stacking on the fixed A3 (`407aa75cc6`) so both the fix **and** the test ride up. Not urgent — the stack is gated behind #11773, which isn't approved yet — flagging so the re-stack isn't forgotten.

🔬 QA Claude

--- a/adapters/repos/db/lsmkv/search_segment.go
+++ b/adapters/repos/db/lsmkv/search_segment.go
@@ -137,11 +137,17 @@ func DoBlockMaxWand(ctx context.Context, limit int, results Terms, averagePropLe
 			// so no explicit exhausted guard is needed here.
 			t := candidates[i]
 			if t.currentBlockMaxId < pivotID {
+				// a real shallow advance moves idPointer without re-sorting (and may
+				// even exhaust mid-array), so the slice may then hold inversions that
+				// only a full sortByID repairs — see the matched branch. blockEntryIdx
+				// changes on every actual move/exhaust but not on the early-return
+				// no-op (docCount==1 terms never initialize currentBlockMaxId, so the
+				// guard over-fires for them; the no-op must not defeat the fast path).
+				prevBlockIdx := t.blockEntryIdx
 				t.AdvanceAtLeastShallow(pivotID)
-				// a shallow advance moves idPointer without re-sorting (and may even
-				// exhaust mid-array), so the slice may now hold inversions that only
-				// a full sortByID repairs — see the matched branch.
-				needFullSort = true
+				if t.blockEntryIdx != prevBlockIdx {
+					needFullSort = true
+				}
 			}
 			upperBound += t.currentBlockImpact
 		}

--- a/adapters/repos/db/lsmkv/search_segment.go
+++ b/adapters/repos/db/lsmkv/search_segment.go
@@ -38,6 +38,15 @@ func DoBlockMaxWand(ctx context.Context, limit int, results Terms, averagePropLe
 	pivotID := uint64(0)
 	var pivotPoint int
 	upperBound := float32(0)
+	// needFullSort tracks whether shallow advances have run since the last full
+	// sortByID. Shallow advances move idPointers (and can exhaust terms) in place,
+	// so the slice may hold inversions or mis-placed exhausted sentinels that only
+	// a full sort repairs. While false, the matched branch can restore order with
+	// a per-advanced-term re-insertion instead of re-sorting all terms — matched
+	// iterations themselves never shallow-advance (their live prefix is already
+	// aligned, so currentBlockMaxId >= pivotID), so matched-dense queries stay on
+	// the cheap path.
+	needFullSort := false
 
 	// a nil ctx is tolerated (some callers pass none): a nil done channel never
 	// selects, so cancellation is simply disabled rather than panicking.
@@ -129,6 +138,10 @@ func DoBlockMaxWand(ctx context.Context, limit int, results Terms, averagePropLe
 			t := candidates[i]
 			if t.currentBlockMaxId < pivotID {
 				t.AdvanceAtLeastShallow(pivotID)
+				// a shallow advance moves idPointer without re-sorting (and may even
+				// exhaust mid-array), so the slice may now hold inversions that only
+				// a full sortByID repairs — see the matched branch.
+				needFullSort = true
 			}
 			upperBound += t.currentBlockImpact
 		}
@@ -162,14 +175,29 @@ func DoBlockMaxWand(ctx context.Context, limit int, results Terms, averagePropLe
 						topKHeap.InsertAndPop(pivotID, score, limit, &worstDist, docInfos)
 					}
 				}
+				advanced := 0
 				for _, term := range results {
 					if !term.exhausted && term.idPointer != pivotID {
 						break
 					}
 					term.Advance()
+					advanced++
 				}
 
-				results.sortByID()
+				if needFullSort {
+					results.sortByID()
+					needFullSort = false
+				} else {
+					// only the advanced prefix moved (each idPointer increased) and the
+					// suffix is inversion-free, so re-inserting the prefix right-to-left
+					// yields the exact permutation a full stable sortByID would: at each
+					// step the suffix is already sorted, and reinsertRight's strict `<`
+					// settles equal ids in original index order, matching insertion-sort
+					// stability.
+					for i := advanced - 1; i >= 0; i-- {
+						results.reinsertRight(i)
+					}
+				}
 
 			} else {
 				nextList := pivotPoint

